packages/ab-testing/src/index.ts

```diff
@@ -25,29 +30,74 @@ function getModuloValue(experiment: string, userId: number | string): number {
     return crc32.calculate(String(userId), crc32.calculate(experiment)) % 100;
 }
 
+function validateAllocationCriteria(filters: ForceInclude, userProfile: UserProfile): boolean {
+    let itemMatchesCriteria = true;
+    for (const key in filters) {
+        if (!filters[key].includes(userProfile[key])) {
+            itemMatchesCriteria = false;
+        }
+    }
+
+    return itemMatchesCriteria;
+}
+
+function validateUserWithinAllocationRange(userSegmentNum: number, range: AllocationV1) {
+    for (const allocation of range) {
+        if (allocation[0] <= userSegmentNum && userSegmentNum < allocation[1]) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+export function validateAllocation(
+    cohort: Cohort,
+    userProfile: UserProfile,
+    configName: string,
+    userId: number | string
+): boolean {
+    const userSegmentNum = getModuloValue(configName, userId);
+    if (cohort.allocation) {
+        if (validateUserWithinAllocationRange(userSegmentNum, cohort.allocation)) {
+            if (cohort.allocation_criteria) {
+                return validateAllocationCriteria(cohort.allocation_criteria, userProfile);
+            }
+            return true;
+        }
+    }
+
+    return false;
+}
+
+function validateForceInclude(cohort: Cohort, userProfile: UserProfile): boolean {
+    if (cohort.force_include) {
+        for (const key in cohort.force_include) {
+            if (cohort.force_include[key].includes(userProfile[key])) {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
 function matchUserCohort(
     experimentConfig: Experiment,
     userId: number | string,
     userProfile: { [s: string]: string }
 ): string {
-    const userSegmentNum = getModuloValue(experimentConfig.name, userId);
     let allocatedCohort = 'control';
     for (const cohort of experimentConfig.cohorts) {
-        if (cohort.force_include) {
-            for (const key in cohort.force_include) {
-                if (cohort.force_include[key].includes(userProfile[key])) {
-                    return cohort.name;
-                }
-            }
+        if (validateForceInclude(cohort, userProfile)) {
+            return cohort.name;
         }
-        if (allocatedCohort === 'control') {
-            for (const allocation of cohort.allocation || []) {
-                if (allocation[0] <= userSegmentNum && userSegmentNum < allocation[1]) {
-                    allocatedCohort = cohort.name;
-                }
-            }
+
+        if (validateAllocation(cohort, userProfile, experimentConfig.name, userId)) {
```

Comment:
we only need to call `validateAllocationCriteria` and `getModuloValue` once for each `matchUserCohort`

Calling them in a loop is wasting CPU cycles. 

Performance is an important goal for this library.

Reply:
Not sure what your point is, but the example looks is valid.

A given user can only match one cohort at a time -- therefore we stop evaluating `allocation`s when there is a match. The code still had to continue the loop on `force_includes` because it meant overriding `allocation` matches.

Reply:
same as https://github.com/appannie/ab-testing/pull/54#discussion_r816173605